### PR TITLE
MOB-20234 - separate event history recording from edge event recording in IAMs

### PR DESCRIPTION
**Pull request by sbenedicadb**
also fix IAM rules engine to properly load historical conditions.

tests to come in the next pr

### CODE REVIEW

File: AEPMessaging/Sources/Message.swift [hunk 1]
```diff
@@ -146,6 +128,44 @@ public class Message: NSObject {
...
+        
+        let mask = [
+            MessagingConstants.Event.History.Mask.EVENT_TYPE,
+            MessagingConstants.Event.History.Mask.MESSAGE_ID,
+            MessagingConstants.Event.History.Mask.TRACKING_ACTION
+        ]
+        
+        let interactionLog = interaction == nil ? "" : " with value '\(interaction ?? "")'"
+        Log.trace(label: MessagingConstants.LOG_TAG, "Writing '\(eventType.propositionEventType)' event\(interactionLog) to EventHistory for in-app message with activityId '\(propInfo.activityId)'")
+        let event = Event(name: "EventHistoryWrite", type: "history", source: "history", data: eventHistoryData, mask: mask)
```
Comment (swarna04) [inline AEPMessaging/Sources/Message.swift:167]: nit: Can we add the name, type, source to the MessagingConstants and avoid hardcoding of the values here.

File: AEPMessaging/Sources/Message.swift [hunk 2]
```diff
@@ -146,6 +128,44 @@ public class Message: NSObject {
...
+        ]
+        
+        let mask = [
+            MessagingConstants.Event.History.Mask.EVENT_TYPE,
+            MessagingConstants.Event.History.Mask.MESSAGE_ID,
+            MessagingConstants.Event.History.Mask.TRACKING_ACTION
+        ]
+        
+        let interactionLog = interaction == nil ? "" : " with value '\(interaction ?? "")'"
+        Log.trace(label: MessagingConstants.LOG_TAG, "Writing '\(eventType.propositionEventType)' event\(interactionLog) to EventHistory for in-app message with activityId '\(propInfo.activityId)'")
```
Comment (swarna04) [inline AEPMessaging/Sources/Message.swift:166]: nit: Can we not just log the info like below?
 ```
 Log.trace(label: MessagingConstants.LOG_TAG, "Writing '\(eventType.propositionEventType)' event with value \(interaction ?? "") to EventHistory for in-app message with activityId '\(propInfo.activityId)'")
 ```
 Or are you trying to discern if empty string is supplied for interaction vs interaction is nil?

**Review by swarna04 — APPROVED** [review]

File: AEPMessaging/Sources/Message+FullscreenMessageDelegate.swift [hunk 3]
```diff
@@ -33,6 +34,7 @@ extension Message: FullscreenMessageDelegate {
         guard let message = message.parent else {
             return
         }
+        message.recordEventHistory(eventType: .dismiss, interaction: nil)
```
Comment (spoorthipujariadobe) [inline AEPMessaging/Sources/Message+FullscreenMessageDelegate.swift:37]: Should we record the IAM history even if showing it failed? If the frequency rule is, say show once, the IAM would never get shown in this case

File: AEPMessaging/Sources/Message+FullscreenMessageDelegate.swift [hunk 4]
```diff
@@ -66,6 +68,7 @@ extension Message: FullscreenMessageDelegate {
             // handle optional tracking
             if let interaction = queryParams[MessagingConstants.IAM.HTML.INTERACTION], !interaction.isEmpty {
                 message?.track(interaction, withEdgeEventType: .interact)
```
Comment (spoorthipujariadobe) [inline AEPMessaging/Sources/Message+FullscreenMessageDelegate.swift:70]: Should probably check for `autoTrack` here and not send the tracking event?

File: AEPMessaging/Sources/Message.swift (hunk 2 above)
Comment (sbenedicadb) [reply AEPMessaging/Sources/Message.swift:166]: there is only a value for `interact` event types. so it looks like the following for each of our 4 types: 
```
Writing 'trigger' event to EventHistory for activityId 'd2b0e969-0378-4878-b669-6741cca0c770#b4ae863b-e517-42da-9af7-897adeff1157'
Writing 'display' event to EventHistory for activityId 'd2b0e969-0378-4878-b669-6741cca0c770#b4ae863b-e517-42da-9af7-897adeff1157'
Writing 'interact' event with value 'clicked' to EventHistory for activityId 'd2b0e969-0378-4878-b669-6741cca0c770#b4ae863b-e517-42da-9af7-897adeff1157'
Writing 'dismiss' event to EventHistory for activityId 'd2b0e969-0378-4878-b669-6741cca0c770#b4ae863b-e517-42da-9af7-897adeff1157'
```

File: AEPMessaging/Sources/Message+FullscreenMessageDelegate.swift (hunk 4 above)
Comment (sbenedicadb) [reply AEPMessaging/Sources/Message+FullscreenMessageDelegate.swift:70]: i don't view this as an auto track - this is an explicit request to track provided by the URL.

happy to discuss though if you think it's confusing or you see it differently.

**Pull request merged**